crates/rpc-ws-server/src/notify_controller.rs

```diff
@@ -0,0 +1,122 @@
+use ckb_channel::{bounded, select, Receiver, RecvError, Sender};
+use ckb_stop_handler::{SignalSender, StopHandler};
+use ckb_types::core::service::Request;
+use gw_types::offchain::ErrorTxReceipt;
+use log::{debug, trace};
+use std::collections::HashMap;
+use std::thread;
+
+pub use ckb_types::core::service::PoolTransactionEntry;
+
+pub const SIGNAL_CHANNEL_SIZE: usize = 1;
+pub const REGISTER_CHANNEL_SIZE: usize = 2;
+pub const NOTIFY_CHANNEL_SIZE: usize = 128;
+
+pub type NotifyRegister<M> = Sender<Request<String, Receiver<M>>>;
+
+#[derive(Clone)]
+pub struct NotifyController {
+    stop: StopHandler<()>,
+
+    new_error_tx_receipt_register: NotifyRegister<ErrorTxReceipt>,
+    new_error_tx_receipt_notifier: Sender<ErrorTxReceipt>,
+}
+
+impl Drop for NotifyController {
+    fn drop(&mut self) {
+        self.stop.try_send(());
+    }
+}
+
+pub struct NotifyService {
+    error_tx_receipt_subscribers: HashMap<String, Sender<ErrorTxReceipt>>,
+}
+
+impl Default for NotifyService {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NotifyService {
+    pub fn new() -> Self {
+        Self {
+            error_tx_receipt_subscribers: HashMap::default(),
+        }
+    }
+
+    /// start background single-threaded service with specified thread_name.
+    pub fn start<S: ToString>(mut self, thread_name: Option<S>) -> NotifyController {
+        let (signal_sender, signal_receiver) = bounded(SIGNAL_CHANNEL_SIZE);
+
+        let (new_error_tx_receipt_registrer, new_error_tx_receipt_register_receiver) =
+            bounded(NOTIFY_CHANNEL_SIZE);
+        let (new_error_tx_receipt_sender, new_error_tx_receipt_receiver) =
+            bounded(NOTIFY_CHANNEL_SIZE);
+
+        let mut thread_builder = thread::Builder::new();
+        if let Some(name) = thread_name {
+            thread_builder = thread_builder.name(name.to_string());
+        }
+        let join_handle = thread_builder
+            .spawn(move || loop {
```

Comment:
We can use smol::spawn instead of start a new thread. The `Sender` and channel should be replaced with smol version

Reply:
Notice: we shouldn't depend on ckb_channel and ckb_stop_handler.